### PR TITLE
[CORE-1712] Add SetProjectDefaults

### PR DESCRIPTION
**Pull request by robert-uhl**

### CODE REVIEW

File: src/pps/pps.proto [hunk 1]
```diff
@@ -1034,4 +1034,7 @@ service API {
 
   // GetProjectDefaults returns the defaults for a particular project.
   rpc GetProjectDefaults(GetProjectDefaultsRequest) returns (GetProjectDefaultsResponse) {}
```
Comment (smalyala) [inline src/pps/pps.proto:1036]: What was the rationale behind putting project-related RPCs in PPS?

**Review by smalyala — APPROVED** [review]
OK on the proto changes

**Comment by robert-uhl** [comment]
One item to mention is that `SetProjectDefaults`’s internals are very similar to `SetClusterDefaults`. My normal rule is to wait until doing something at least three times before abstracting it out, but I am very open to abstracting this out now if @acohen4 concurs.

File: src/server/pps/server/defaults_test.go [hunk 2]
```diff
@@ -246,6 +246,93 @@ func TestAPIServer_CreatePipelineV2_regenerate(t *testing.T) {
 	require.Equal(t, false, req.Autoscaling, "autoscaling is still false")
 }
 
+func TestAPIServer_SetProjectDefaults_regenerate(t *testing.T) {
+	ctx := pctx.TestContext(t)
```
Comment (acohen4) [inline src/server/pps/server/defaults_test.go:250]: WDYT about having a test that lists out test cases as structs of {"field", "ClusterDefault", "ProjectDefault", "SetValue", "regenerate"}, and have expected final value for the field?

File: src/server/pps/server/api_server.go [hunk 3]
```diff
@@ -3959,20 +3959,100 @@ func (a *apiServer) GetProjectDefaults(ctx context.Context, req *pps.GetProjectD
...
+	if req.Regenerate {
+		// Determine if the new defaults imply changes to any pipelines.
+		// To do this, each pipeline’s info is synthesized to a
+		// CreatePipelineRequest using the same logic already ultimately
+		// used by `pachctl edit pipeline`, then that is turned into
+		// JSON and merged with the new defaults.  The result of the
+		// merger is then unmarshalled into a CreatePipelineRequest and
+		// equality is checked with proto.Equal.
+		pp = make(map[*pps.Pipeline]*pps.CreatePipelineTransaction)
+		if err := a.listPipeline(ctx, &pps.ListPipelineRequest{Details: true}, func(pi *pps.PipelineInfo) error {
```
Comment (acohen4) [inline src/server/pps/server/api_server.go:3997]: I don't think ListPipelineRequest.Details is being interpreted anywhere in the system

File: src/server/pps/server/api_server.go [hunk 4]
```diff
@@ -3959,20 +3959,100 @@ func (a *apiServer) GetProjectDefaults(ctx context.Context, req *pps.GetProjectD
...
+			if !proto.Equal(pi.GetPipeline().GetProject(), req.GetProject()) {
+				fmt.Println("returning")
```
Comment (acohen4) [inline src/server/pps/server/api_server.go:3999]: remove println

Comment (robert-uhl) [reply src/server/pps/server/api_server.go:3999]: Argh, good catch.

File: src/server/pps/server/defaults_test.go (hunk 2 above)
Comment (robert-uhl) [reply src/server/pps/server/defaults_test.go:250]: It’s an intriguing idea. I think one of the key observables is the sequence of steps. Can this be deferred to CORE-1723?

Comment (acohen4) [reply src/server/pps/server/defaults_test.go:250]: yeah totally. A suggestion

**Review by acohen4 — APPROVED** [review]
LGTM

**Pull request merged**